src/AspNetCore/Src/aspnetcoreapplication.cxx

```diff
@@ -0,0 +1,516 @@
+#include "precomp.hxx"
+#include "fx_ver.h"
+#include <algorithm>
+
+typedef int(*hostfxr_main_fn) (const int argc, const wchar_t* argv[]);
+
+// Initialization export
+
+extern "C" __declspec(dllexport)
+VOID
+register_callbacks(
+    _In_ PFN_REQUEST_HANDLER request_handler,
+    _In_ PFN_SHUTDOWN_HANDLER shutdown_handler,
+    _In_ VOID* pvRequstHandlerContext,
+    _In_ VOID* pvShutdownHandlerContext
+)
+{
+    ASPNETCORE_APPLICATION::GetInstance()->SetCallbackHandles(request_handler, shutdown_handler, pvRequstHandlerContext, pvShutdownHandlerContext);
+}
+
+extern "C" __declspec(dllexport)
+HTTP_REQUEST*
+http_get_raw_request(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->GetRequest()->GetRawHttpRequest();
+}
+
+extern "C" __declspec(dllexport)
+HTTP_RESPONSE*
+http_get_raw_response(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->GetResponse()->GetRawHttpResponse();
+}
+
+extern "C" __declspec(dllexport) VOID http_set_response_status_code(
+    _In_ IHttpContext* pHttpContext,
+    _In_ USHORT statusCode,
+    _In_ PCSTR pszReason
+)
+{
+    pHttpContext->GetResponse()->SetStatus(statusCode, pszReason);
+}
+
+extern "C" __declspec(dllexport)
+HRESULT
+http_post_completion(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->PostCompletion(0);
+}
+
+extern "C" __declspec(dllexport)
+VOID
+http_indicate_completion(
+    _In_ IHttpContext* pHttpContext,
+    _In_ REQUEST_NOTIFICATION_STATUS notificationStatus
+)
+{
+    pHttpContext->IndicateCompletion(notificationStatus);
+}
+
+extern "C" __declspec(dllexport)
+VOID
+http_get_completion_info(
+    _In_ IHttpCompletionInfo2* info,
+    _Out_ DWORD* cbBytes,
+    _Out_ HRESULT* hr
+)
+{
+    *cbBytes = info->GetCompletionBytes();
+    *hr = info->GetCompletionStatus();
+}
+
+extern "C" __declspec(dllexport)
+BSTR // TODO probably should make this a wide string
```

Comment:
why do you want to return BSTR a copy of the memory? Is it required by managed layer? or some tried to change the value?

Reply:
The value is required for the managed layer to know where the content root is. Before I had this returned as an out parameter, however I was running into Heap Corruption (I think both managed and native were freeing). @davidfowl did you manage to follow up with whoever is knowledgeable about this?